### PR TITLE
resource/aws_lambda_function: recompute versioned attributes on config update

### PR DESCRIPTION
**Pull request by nemreid**
Add conditional to the `updateComputedAttributesOnPublish` function to recompute attributes on config-only updates. This section of the code was missed in #11211, which enabled config-only versioning. It means that the provider does not recompute the `qualified_arn` and `version` attributes after publishing config-only version updates, which breaks the graph for any downstream resources dependent on those attributes, requiring a refresh and reapply.

Confirmed to recompute outputs correctly in ad-hoc testing:

```
~ qualified_arn                  = "arn:aws:lambda:(region):(account-id):function:MyLambdaFunction:4" -> (known after apply)
```

Apologies for missing this in the original PR to @bflad and the community.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_lambda_function: fix bug where qualified_arn/version attributes were not updated on config-only updates
```

Output from acceptance testing:

```
N/A (did not run)
```

### CODE REVIEW

File: aws/resource_aws_lambda_function.go [hunk 1]
```diff
@@ -685,101 +689,87 @@ func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) e
...
+					variables := readEnvironmentVariables(environmentVariables.(map[string]interface{}))
 
+					configReq.Environment = &lambda.Environment{
+						Variables: aws.StringMap(variables),
+					}
+				}
+			} else {
 				configReq.Environment = &lambda.Environment{
-					Variables: aws.StringMap(variables),
+					Variables: aws.StringMap(map[string]string{}),
```
Comment (nemreid) [inline aws/resource_aws_lambda_function.go]: This interesting diff is actually just me indenting the shown lines to nest within the conditional block.

File: aws/resource_aws_lambda_function.go [hunk 2]
```diff
@@ -665,6 +665,10 @@ func needsFunctionCodeUpdate(d resourceDiffer) bool {
 	return d.HasChange("filename") || d.HasChange("source_code_hash") || d.HasChange("s3_bucket") || d.HasChange("s3_key") || d.HasChange("s3_object_version")
 }
 
+func needsFunctionConfigUpdate(d resourceDiffer) bool {
+	return d.HasChange("description") || d.HasChange("handler") || d.HasChange("memory_size") || d.HasChange("role") || d.HasChange("timeout") || d.HasChange("kms_key_arn") || d.HasChange("layers") || d.HasChange("dead_letter_config") || d.HasChange("tracing_config") || d.HasChange("vpc_config") || d.HasChange("runtime") || d.HasChange("environment")
```
Comment (nemreid) [inline aws/resource_aws_lambda_function.go]: Downside to this PR is checking each of these fields twice worst case, as the existing logic still checks each individual field, although this part short-circuits in most cases.

File: aws/resource_aws_lambda_function.go [hunk 3]
```diff
@@ -267,7 +267,7 @@ func resourceAwsLambdaFunction() *schema.Resource {
 }
 
 func updateComputedAttributesOnPublish(d *schema.ResourceDiff, meta interface{}) error {
-	if needsFunctionCodeUpdate(d) {
+	if needsFunctionCodeUpdate(d) || needsFunctionConfigUpdate(d) {
```
Comment (nemreid) [inline aws/resource_aws_lambda_function.go]: Crucially, this was the function where an update was missed in #11211.

**Comment by nemreid** [comment]
I've narrowed it down to `d.HasChange("vpc_config")` which appears to **always** return true, regardless of whether the values are provided through variables or hardcoded into my `main.tf`.

Testing confirms VPC config **is** version-specific, so recalculating the version on a VPC is necessary. 

I'm wondering if there is something funky with the way `d.HasChange("vpc_config")` confirms or whether I should update the `DiffSuppressFunc` section of `resourceAwsLambdaFunction` that deals with VPCs in order to alleviate this.

Any advice would be appreciated.

**Comment by andresvia** [comment]
@nemreid , thanks for this PR, I'm new to this and I was wondering if this PR will also help with  #12266

As for what `d.HasChange("vpc_config")` is actually doing, would need to check, I thought changes could/should be calculated based on the config+statefile it seems like I'm wrong, here the change (`HasChange`) is calculated based on config+statefile+responses from the service?

Thanks.

**Comment by nemreid** [comment]
This issue may have been resolved by https://github.com/hashicorp/terraform-provider-aws/pull/15121. I will do some testing when possible, and if resolved I will close this PR.

**Comment by nemreid** [comment]
After performing testing, I can confirm that I am seeing the desired behavior on config-only updates such as only changing the function `timeout` or `environment` on versioned lambdas, which does result in a new version number. This PR is no longer necessary so I am closing it.

**Pull request closed without merging**